templates/src/csharp/Grpc.Core/Internal/NativeMethods.Generated.cs.template

```diff
@@ -0,0 +1,233 @@
+      'void grpcsharp_auth_context_release(IntPtr authContext)',
+      '',
+      'Timespec gprsharp_now(ClockType clockType)',
+      'Timespec gprsharp_inf_future(ClockType clockType)',
+      'Timespec gprsharp_inf_past(ClockType clockType)',
+      '',
+      'Timespec gprsharp_convert_clock_type(Timespec t, ClockType targetClock)',
+      'int gprsharp_sizeof_timespec()',
+      '',
+      'CallError grpcsharp_test_callback([MarshalAs(UnmanagedType.FunctionPtr)] NativeCallbackTestDelegate callback)',
+      'IntPtr grpcsharp_test_nop(IntPtr ptr)',
+      'void grpcsharp_test_override_method(string methodName, string variant)',
+    ]
+    
+    import re
+    native_methods = []
+    for signature in native_method_signatures:
+      if not signature:
+        continue
+      match = re.match('([^ ]+) +([A-Za-z0-9_]+)\\((.*)\\)(.*)', signature)
+      native_methods.append({'name': match.group(2), 'returntype': match.group(1), 'params': match.group(3), 'comment': match.group(4)})
+    
+  
+    #native_methods = [
+    #  {'name': 'grpcsharp_init', 'returntype': 'void', 'params': ''},
+    #  {'name': 'grpcsharp_server_request_call', 'returntype': 'CallError', 'params': 'ServerSafeHandle server, CompletionQueueSafeHandle cq, RequestCallContextSafeHandle ctx'},
+    #]
+  %>
+  #region Copyright notice and license
+  
+  // Copyright 2015 gRPC authors.
+  //
+  // Licensed under the Apache License, Version 2.0 (the "License");
+  // you may not use this file except in compliance with the License.
+  // You may obtain a copy of the License at
+  //
+  //     http://www.apache.org/licenses/LICENSE-2.0
+  //
+  // Unless required by applicable law or agreed to in writing, software
+  // distributed under the License is distributed on an "AS IS" BASIS,
+  // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  // See the License for the specific language governing permissions and
+  // limitations under the License.
+  
+  #endregion
+  
+  using System;
+  using System.Collections.Concurrent;
+  using System.Diagnostics;
+  using System.IO;
+  using System.Reflection;
+  using System.Runtime.InteropServices;
+  using System.Threading;
+  
+  using Grpc.Core.Logging;
+  using Grpc.Core.Utils;
+  
+  namespace Grpc.Core.Internal
+  {
+      internal partial class NativeMethods
+      {
+          #region Native methods
+          
+          % for method in native_methods:
+          public readonly Delegates.${method['name']}_delegate ${method['name']};
+          % endfor
+
+          #endregion
+  
+          public NativeMethods(UnmanagedLibrary library)
+          {
+              % for method in native_methods:
+              this.${method['name']} = GetMethodDelegate<Delegates.${method['name']}_delegate>(library);
+              % endfor
+          }
+          
+          public NativeMethods(DllImportsFromStaticLib unusedInstance)
```

Comment:
nit: can we use different method names to differentiate the different types of `NativeMethod` instantion, rather than differing signatures? Perhaps use a some sort of `NativeMethod` factory?

Reply:
I was trying to avoid this, but if you wanted to have a factory method, you need a constructor anyway to inititialize all the readonly fields - so having a factory methods would only add a lot of generated code (and we would still need to have some way to disambiguate the constructors).

IMHO What we have is the best I could do at the moment and it's internal code (and despite the small hack I'd say it's sufficiently clear). But I'm open to better ideas ;-)

Reply:
Yeah this is definitely a minor issue :) but one other idea as an optional suggestion:

instead of `class type`s, perhaps`DllImportsFromSharedLib` and `DllImportsFromSharedLib` could be constant variables of class `DllImportsFromLib`?
 Maybe `DllImportsFromLib` could then take a constructor param for the `ImportName` field and there could be a `NativeMethods(DllImportsFromLib)`that could be passed those instances, which might de-dupe a few lines from the template code.

Reply:
That wouldn't work because the methods with [DllImport] attribute needs to be static - so passing anything to the constructor won't help. I think the duplication here can't really be avoided (unless we want to use reflection to set all the delegates), which is also why I decided to use codegeneration here.